### ISSUE TITLE
Allow non-expiry of backups

### ISSUE THREAD

**Issue by stijzermans** [vmware-tanzu/velero#8099]
**Describe the problem/challenge you have**
Backups are now forced to have an expiry - where technically the maximum value of the duration of this value is approx. 290 years (https://pkg.go.dev/time#Duration)  or `2562047h`, this leads to unneeded hourly evaluations during garbage collection and is from a consistency point of view not the same as a non-expiring backup.

**Describe the solution you'd like**
An explicit TTL value of `0` or `-1` is processed as non-expiry an hence will not set the expiry attribute of the backup.
If the TTL parameter is omitted,  the `defaultBackupTTL` is still used and respected.

**Vote on this issue!**
This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

**Comment by sseago** [comment]
I don't have a strong opinion on whether this would be a good thing to add, but a quick note that the suggestion of `0` as a non-expiry value won't work -- since ttl is not a pointer field, setting `0` would be equivalent to omitting the param. Currently it looks like if you specify a negative duration, velero just accepts it and would set expiration to some time *before* now, meaning it's always expired. It would be possible to modify this to *not* set expiration if ttl is negative.

The gc controller would still evaluate the backup for expiration, but it wouldn't need to calculate time between now and 290 years from now, since we already check for empty expiration:
```
        if backup.Status.Expiration == nil || backup.Status.Expiration.After(now) {
                log.Debug("Backup has not expired yet, skipping")
                return ctrl.Result{}, nil
        }
```

**Comment by reasonerjt** [comment]
This maybe do-able, but if a backup is old enough it may not be restored by the recent versions of velero.
IMO create backups regularly is a better practice.

**Comment by Lirt** [comment]
Hello, 

this enhancement request is an outcome of incremental backup implementation for Openstack provider. 

We discussed the relevance of backup cleanup in case of incremental backup scenario. There it doesn't make sense to expire the backup as the first backup is full and removing it will degrade all other backups and in the same way removing one of the incremental backups will degrade all backups made after it. 

So in case of incremental backups we would like to have a schedule that creates new backups, but doesn't clean up or remove old|expired ones. 

Customer|User is able to create new incremental backup anytime he/she wants and remove the old one.

Let me know if this enhancement makes sense after the explanation.

Is there any other Velero backup provider, that offers incremental backups? How is this situation with schedule solved there?

**Comment by sseago** [comment]
@Lirt So the only part of the backup that's incremental is the volume data -- the kubernetes metadata is always a full backup, and the backup tarball from an earlier backup is never used if you are restoring a newer backup. The part that is incremental is in the volume data. For fs-backup, datamover, and the officially maintained volume snapshotter plugins, incremental volume backup is based on retaining previous snapshots that current incremental snapshots depend on, so if I create backup1 with full volume backups, and then backup2 does incremental volume backups, when backup1 expires, it has no effect on how backup2 is restored, since any kopia or restic snapshots from backup1 that are needed for backup2's volumes will be left intact until backup2 is deleted. The aws/azure/gcp volume snapshotters work similarly.

Is there some aspect of the velero-plugin-for-openstack volume snapshotter plugins that relies on prior Backup objects to exist and not just prior snapshots?

**Comment by Lirt** [comment]
Hi @sseago,

Yes you are right about metadata backup and volume snapshots / backups.

Sorry for late reply. In your answer you are referring to a restic/kopia backup. The backup/snapshot plugins are independent of the restic/kopia backups (users can use them or skip them). The volume snapshotter plugins are here to make snapshot or variations of a snapshot. In case of Openstack there is a support for backups which you can use instead of snapshots.

This backup creation is handled by Openstack component (cinder). The use-case supported by API is that you can create one full backup and then incremental backups. So the difference in velero-plugin-for-openstack is that instead of a snapshot (non-durable) it can create backup (durable) and thus it makes sense to skip restic/kopia backup action (because the backup component ensures the backup is durable (up to cloud provider how to do it) and there is no need to run them.

In this scenario the user wants to create first full backup and then incremental backups and then expiration doesn't make sense, because deletion of intermediate backups of the full backup will corrupt the backup chain.

**Comment by sseago** [comment]
Hmm. That sounds like it might be a bug in the openstack volume snapshotter or cinder back end. For example, ebs snapshots are also incremental, so if I take snapshot1, then snapshot2, I'm pretty sure that on the ebs back end, the dependency is managed -- so if I delete the velero backup1, the backup2 with incremental snapshot2 should still work. Does the cinder infrastructure not track what snapshot an incremental snapshot depends on (and thus prevent its deletion as long as there is a newer incremental backup that depends on it?

**Comment by Lirt** [comment]
Yes, Cinder tracks the dependency automatically.  So deletion of a backup that has a dependent backup is impossible. Therefore in our scenario we are basically able to only remove newest backup (which has no dependants).

Expired TTL starts to remove from the oldest backup, so every backup deletion will fail (because when deleting from oldest each backup will have a dependent backup). That's why we want to be able to not configure TTL at all.
